### ISSUE TITLE
Buildbot uploads R1/beta1 repositories to master S3 bucket

### ISSUE THREAD

**Issue by waddlesplash** [haiku/infrastructure#43]
See last step of: https://build.haiku-os.org/buildbot/#/builders/28/builds/333

This was a R1/beta1 repo build, but it got shoved into the "master" S3 bucket.

**Comment by waddlesplash** [comment]
Manually moved the directories so the updates are now in place, anyway.

**Comment by kallisti5** [comment]
This is likely due to:
https://github.com/haiku/infrastructure/commit/c1f121d5015e3161503f38576e49e663921f96d9#diff-d3c5ee9e6be39d2b99a1af34e4c0a2d9

Adding another codebase, we reference "branch" during the built.  Previously there was one codebase in the scheduler with one branch... now we get two branches in two codebases.  I think we're grabbing the new codebase (last item in array).

This can be seen in the debug output for the build referenced above...

![1](https://user-images.githubusercontent.com/100110/59440291-53c07280-8dbc-11e9-8ba1-8d7c7355d7a2.png)

**Comment by kallisti5** [comment]
Raised a bug to buildbot here:
https://github.com/buildbot/buildbot/issues/4831

It looks as though introducing another codebase silently unsets the 'branch' property. We then use the default which is 'master'.

Asking the buildbot folks how you're supposed to determine the branch in a multi-codebase build.

**Comment by kallisti5** [comment]
Pushe

**Comment by kallisti5** [comment]
ok.. So when you have multiple codebases, buildbot expects you to leverage "Interpolate"
After much struggling, "Interpolate" is pretty much an attempt to hack futures into python. (something defined, that is set at execution)

The issue is we lose access to the Interpolated data pretty quickly.

Here's an example of trying to leverage Interpolate:
```
branch = Interpolate("%(src:buildtools:branch)s")
crossDir = "../cross-tools-%s-%s" % (branch, targetArch)
```
In the example above, crossDir is set to "../cross-tools-Interpolate('%(src:buildtools:branch)s')-x86"

This is because the Interpolate value is only set on execution of the job.

In another example:
```
branch = Interpolate("%(src:buildtools:branch)s")
if branch == "master":
  blah
```

this also doesn't work, since branch is "Interpolate("%(src:buildtools:branch)s")" until execution.
The whole Interpolate model isn't well suited for all the business logic we need in Python.  I got pretty close replacing *everything* that uses an Interpolated value with something like this:

```
crossToolsDir = Interpolate("../cross-tools-%%(src:buildtools:branch)s-%s" % targetArch)
gccPath = Interpolate("../cross-tools-%%(src:buildtools:branch)s-%s/cross-tools-%s/bin/gcc-%s" % (targetArch, targetArch, machineTriplet))
```

That code smells pretty bad however, and it doesn't address the lack of "if brach" within the codebase.

**Comment by kallisti5** [comment]
honestly... at this point I think the "correct" solution is to revert this:
c1f121d5015e3161503f38576e49e663921f96d9

While that commit is "correct", the buildbot Interpolate crap adds a *lot* of headaches and makes some pretty hard to follow code.

**Comment by waddlesplash** [comment]
Can't we just use shell wizardry to get the branch from Git directly, instead of from buildbot vars, when uploading to Minio?

**Comment by kallisti5** [comment]
Damn you and your good ideas.

**Comment by kallisti5** [comment]
errr.. a *LOT* of magic logic depends on the branch name within the python code. Pretty much all of our build profile choices are formed in the python code based on the current branch. (none of this is my doing btw :-) )

Maybe the real solution to this is finishing my work using concourse.ci :-D

**Comment by kallisti5** [comment]
Here's the most complex code path that needs updated:

* We write the jam command profiles out like %%PROFILE%% per arch:
  * https://github.com/haiku/infrastructure/blob/master/data/buildbot/master.cfg#L402
* We pick and choose the profile commands we want here into a property:
  * https://github.com/haiku/infrastructure/blob/master/data/buildbot/master.cfg#L1217
* We then grab the list of profiles at build, and sub in the actual profiles names based on branch:
  * https://github.com/haiku/infrastructure/blob/master/data/buildbot/master.cfg#L871

Since fixing this will require  a *massive* refactor of our buildbot code... i'd rather just revert the (correct, but massively incomplete) c1f121d

**Comment by kallisti5** [comment]
there is one alternative... we move the scripts to build releases into our git repository.

The script will get all the important bits (or read them from BuildConfig), and then produce a release based on the inputs.

buildHaikuRelease arm
buildHaikuRepository arm

We could also roll the hacked UEFI scripts for r1beta1 into these scripts as well.

**Comment by kallisti5** [comment]
buildbot is no more.
